### ISSUE TITLE
Question about weight normalization

### ISSUE THREAD

**Issue by j0ntt** [maddin79/darch#10]
Hi,

I have another question regarding the weight normalization. The manual says 
`normalizeWeights` `Logical indicating whether to normalize weights (L2 norm = 1).`

So does this mean that each weight-matrix between layers is normalized this way, and not all weight matrices together so that their `L2 norm = 1`? Is there some reference to it or how did you come up with it?

Thank you again.

**Comment by saviola777** [comment]
Hi,

It is similar to max norm regularization described [here](https://cs231n.github.io/neural-networks-2/#reg), the only difference being that it does not enforce an upper bound, but enforce a specific L2 norm (1) for each incoming weight vector. My goal was to prevent weights from getting out of hand and causing numeric problems and this was the first thing I came up with, probably not the best idea. The current git version has weight normalization much closer to the implementation described in the first section of appendix A in [this](http://arxiv.org/abs/1207.0580) publication, using a max norm with default bound 15.

**Comment by j0ntt** [comment]
Ok good. It seems to help. At least with the basic backpropagation. Weight-decay is another regularization method for that purpose. So each weight vector is normalized separately?

**Comment by saviola777** [comment]
Exactly, here's the code:

```R
# normalize weights
if (darch@normalizeWeights)
{
  value[1:(nrow(value) - 1),] <-
    apply(value[1:(nrow(value) - 1),, drop = F], 2,
          function (e) { e/norm(e, "2") })
}
```

Where ```value``` is one weight matrix (the bias values are excluded from the normalization). The problem with weight decay is that weights tend to grow faster than they decay, so it doesn't really replace normalization.

**Comment by saviola777** [comment]
Closing after a month of inactivity, open a new issue if something related comes up.
